src/manager.rs

```diff
@@ -113,33 +111,32 @@ impl<S: MutEventSubscriber> EventManager<S> {
     ///
     /// On success, it returns number of dispatched events or 0 when interrupted by a signal.
     pub fn run_with_timeout(&mut self, milliseconds: i32) -> Result<usize> {
-        let event_count = match self.epoll_context.epoll.wait(
-            self.ready_events.len(),
-            milliseconds,
-            &mut self.ready_events[..],
-        ) {
-            Ok(ev) => ev,
-            // EINTR is not actually an error that needs to be handled. The documentation
-            // for epoll.run specifies that run exits when it for an event, on timeout, or
-            // on interrupt.
-            Err(e) if e.raw_os_error() == Some(libc::EINTR) => return Ok(0),
-            Err(e) => return Err(Error::Epoll(Errno::from(e))),
-        };
+        let event_count = self.epoll_context.poll(milliseconds)?;
         self.dispatch_events(event_count);
 
         Ok(event_count)
     }
 
     fn dispatch_events(&mut self, event_count: usize) {
+        // EpollEvent doesn't implement Eq or PartialEq, so...
+        let default_event: EpollEvent = EpollEvent::default();
+
         // Used to record whether there's an endpoint event that needs to be handled.
         #[cfg(feature = "remote_endpoint")]
         let mut endpoint_event = None;
 
-        // Use the temporary, pre-allocated buffer to check ready events.
+        // TODO: implement an iterator for EpollWrapper to simplify the abstraction
+        // https://github.com/rust-vmm/event-manager/issues/44
         for ev_index in 0..event_count {
-            let event = self.ready_events[ev_index];
+            let event = self.epoll_context.ready_events[ev_index];
             let fd = event.fd();
 
+            // Check whether this event has been discarded.
+            // EpollWrapper::remove_event() discards an IO event by setting it to default value.
+            if event.events() == default_event.events() && fd == default_event.fd() {
```

Comment:
How about we only check `if event.events() == default_event.events()`? The default fd value is actually valid (0) anyway. Going forward we should add some accessor to `EpollEvent` so we can set the fd to something like `-1` to invalidate the event, and we can also detect invalidated events on the slower path where we end up if the fd is not present in the `fd_dispatch` map.

Reply:
@alexandruag can you add an issue with this as well?

Reply:
Yes, I have tried to change the vmm-sys-utils to use -1 as default value for fds, but that needs synchronization with these two crates. And I guess "event.events() == default_event.events()" is enough, but haven't checked the kernel implementation yet.

Reply:
It's not really enough I think because the fd you're checking from `default_event` does not necessarily needs to be the fd within the events.